File: hackerrank/matrix/diagdiff1.js

// Solution to https://www.hackerrank.com/challenges/diagonal-difference/problem

'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', function() {
    inputString = inputString.replace(/\s*$/, '')
        .split('\n')
        .map(str => str.replace(/\s*$/, ''));

    main();
});

function readLine() {
    return inputString[currentLine++];
}

/**
  * @brief The O(N) solution JS implementation is very similar to the C++ one 
  * @note The arr is an Array of Arrays which is the typical Core Language (with no external libs) representation for a Matrix and an element can be accessed with double square bracket syntax 
  * @note The Array Size is like in C++ represented as a Container property, just a syntactic difference as C++ provides it via the size() member while JS provides it via the length property 
  */
function diagonalDifference(arr) 
{
    let d1=0; 
    let d2=0; 
    for(let i=0; i<arr.length; ++i)
    {
        d1 += arr[i][i]; 
        d2 += arr[i][arr.length-1-i];
    }
    return Math.abs(d1-d2); 
}

function main() {
    const ws = fs.createWriteStream(process.env.OUTPUT_PATH);

    const n = parseInt(readLine(), 10);

    let arr = Array(n);

    for (let i = 0; i < n; i++) {
        arr[i] = readLine().split(' ').map(arrTemp => parseInt(arrTemp, 10));
    }

    const result = diagonalDifference(arr);

    ws.write(result + '\n');

    ws.end();
}

File: hackerrank/matrix/diagdiff1.scala

// Solution to https://www.hackerrank.com/challenges/diagonal-difference/problem

import java.io._
import java.math._
import java.security._
import java.text._
import java.util._
import java.util.concurrent._
import java.util.function._
import java.util.regex._
import java.util.stream._

object Solution {

    // Complete the diagonalDifference function below.
    def diagonalDifference(arr: Array[Array[Int]]): Int = 
    {
        var d1: Int = 0 
        var d2: Int = 0 
        for(i <- 0 to arr.size-1)
        {
            d1 += arr(i)(i)
            d2 += arr(i)(arr.size-1-i)
        }
        return Math.abs(d1-d2)
    }

    def main(args: Array[String]) {
        val stdin = scala.io.StdIn

        val printWriter = new PrintWriter(sys.env("OUTPUT_PATH"))

        val n = stdin.readLine.trim.toInt

        val arr = Array.ofDim[Int](n, n)

        for (i <- 0 until n) {
            arr(i) = stdin.readLine.split(" ").map(_.trim.toInt)
        }

        val result = diagonalDifference(arr)

        printWriter.println(result)

        printWriter.close()
    }
}


